subsys/bluetooth/audio/aics_client.c

```diff
@@ -0,0 +1,889 @@
+#define LOG_MODULE_NAME bt_aics_client
+#include "common/log.h"
+
+static struct bt_aics aics_insts[CONFIG_BT_MAX_CONN * CONFIG_BT_AICS_CLIENT_MAX_INSTANCE_COUNT];
+
+static int aics_client_common_control(struct bt_conn *conn, uint8_t opcode, struct bt_aics *inst);
+
+static struct bt_aics *lookup_aics_by_handle(struct bt_conn *conn, uint16_t handle)
+{
+	for (int i = 0; i < ARRAY_SIZE(aics_insts); i++) {
+		if (aics_insts[i].cli.conn == conn &&
+		    aics_insts[i].cli.active &&
+		    aics_insts[i].cli.start_handle <= handle &&
+		    aics_insts[i].cli.end_handle >= handle) {
+			return &aics_insts[i];
+		}
+	}
+
+	BT_DBG("Could not find AICS instance with handle 0x%04x", handle);
+
+	return NULL;
+}
+
+uint8_t aics_client_notify_handler(struct bt_conn *conn, struct bt_gatt_subscribe_params *params,
+				   const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct bt_aics *inst = lookup_aics_by_handle(conn, handle);
+	struct bt_aics_state *state;
+	uint8_t *status;
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (!data || !length) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	if (handle == inst->cli.state_handle) {
+		if (length == sizeof(*state)) {
+			state = (struct bt_aics_state *)data;
+			BT_DBG("Inst %p: Gain %d, mute %u, gain_mode %u, counter %u",
+			       inst, state->gain, state->mute, state->gain_mode,
+			       state->change_counter);
+
+			inst->cli.change_counter = state->change_counter;
+
+			if (inst->cli.cb && inst->cli.cb->state) {
+				inst->cli.cb->state(conn, inst, 0, state->gain,
+						    state->mute, state->gain_mode);
+			}
+		}
+	} else if (handle == inst->cli.status_handle) {
+		if (length == sizeof(*status)) {
+			status = (uint8_t *)data;
+			BT_DBG("Inst %p: Status %u", inst, *status);
+			if (inst->cli.cb && inst->cli.cb->status) {
+				inst->cli.cb->status(conn, inst, 0, *status);
+			}
+		}
+	} else if (handle == inst->cli.desc_handle) {
+		char desc[MIN(CONFIG_BT_L2CAP_RX_MTU, BT_ATT_MAX_ATTRIBUTE_LEN) + 1];
+
+		/* Truncate if too large */
+		length = MIN(sizeof(desc) - 1, length);
+
+		memcpy(desc, data, length);
+		desc[length] = '\0';
+		BT_DBG("Inst %p: Input description: %s", inst, log_strdup(desc));
+		if (inst->cli.cb && inst->cli.cb->description) {
+			inst->cli.cb->description(conn, inst, 0, desc);
+		}
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t aics_client_read_state_cb(struct bt_conn *conn, uint8_t err,
+					 struct bt_gatt_read_params *params,
+					 const void *data, uint16_t length)
+{
+	int cb_err = err;
+	struct bt_aics *inst = lookup_aics_by_handle(conn, params->single.handle);
+	struct bt_aics_state *state = (struct bt_aics_state *)data;
+
+	memset(params, 0, sizeof(*params));
+
+	if (!inst) {
+		BT_DBG("Instance not found");
+		return BT_GATT_ITER_STOP;
+	}
+
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+	inst->cli.busy = false;
+
+	if (data) {
```

Comment:
Here, the data are used even if there is an error (err is set).
Is that wise?  
I am not sure if we have a common pattern, but we at least also have examples of the opposite.

(Also other places.)

Reply:
That is not wise. I can't remember why this was removed, but certainly shouldn't have been. Added.